hw_1.2/hm_1.2/Program.cs

```diff
@@ -0,0 +1,198 @@
+            int[,] circlePermutations = new int[strInASCII.Length, strInASCII.Length];
+
+            string result = "";
+            int numberOfoOriginalString = 1;
+            bool isOriginalStringSorted = false;
+
+            //Creat all circle permutations
+            for (int i = 0; i < strInASCII.Length; ++i)
+            {
+                for (int j = 0; j < strInASCII.Length; ++j)
+                {
+                    circlePermutations[i, j] = strInASCII[(i + j) % strInASCII.Length];
+                }
+            }
+            // step by step find the smallest permutation and write last char in result
+            for (int i = 0; i < strInASCII.Length; ++i)
+            {
+
+                int min = 0;
+                for (int j = 0; j < strInASCII.Length; ++j)
+                {
+                    int depth = 0;
+                    while (depth < strInASCII.Length && circlePermutations[min, depth] == circlePermutations[j, depth])
+                    {
+                        ++depth;
+                    }
+                    if (depth < strInASCII.Length && circlePermutations[min, depth] > circlePermutations[j, depth])
+                    {
+                        min = j;
+                    }
+
+                }
+                // next two conditions are needed to find number of original permutation
+                if (min == 0)
+                {
+                    isOriginalStringSorted = true;
+                }
+
+                if (!isOriginalStringSorted)
+                {
+                    ++numberOfoOriginalString;
+                }
+                result += (char)circlePermutations[min, strInASCII.Length - 1];
+                circlePermutations[min, 0] = 1000;
+            }
+
+
+            return result + " " + Convert.ToString(numberOfoOriginalString);
+        }
+
+        static string ReverseBTW(string str, int numberOfOriginString)
+        {
+            --numberOfOriginString;
+
+            // form alphabet
+            string alphabet = "";
+            for (int i = 0; i < 127; ++i)
+            {
+                if (str.Contains((char)i))
+                {
+                    alphabet += Convert.ToString((char)i);
+                }
+            }
+
+            // amount of every sign
+            int[] count = new int[alphabet.Length];
+            for (int i = 0; i < count.Length; ++i)
+            {
+                for (int j = 0; j < str.Length; ++j)
+                {
+                    if (str[j] == alphabet[i])
+                    {
+                        ++count[i];
+                    }
+                }
+            }
+
+            int[] sum = new int[count.Length];
+            sum[0] = 1;
+            for (int i = 1; i < sum.Length; ++i)
+            {
+                sum[i] = count[i - 1] + sum[i - 1];
+            }
+
+            // form auxiliary array
+            int[] p = new int[str.Length];
```

Comment:
```suggestion
            var p = new int[str.Length];
```
Однобуквенное имя переменной - плохо.